shared/x509utils/types.go

```diff
@@ -27,3 +30,40 @@ type PrivateKey interface {
 type PublicKey interface {
 	Equal(x crypto.PublicKey) bool
 }
+
+// BlockToPrivateKey parses a pem Block looking for rsa, ecdsa or ed25519 Private Keys
+func BlockToPrivateKey(block *pem.Block) (PrivateKey, error) {
+	if block.Type == "PRIVATE KEY" || strings.HasSuffix(block.Type, " PRIVATE KEY") {
+		if pk, _ := x509.ParsePKCS1PrivateKey(block.Bytes); pk != nil {
+			return pk, nil
+		} else if pk, err := x509.ParsePKCS8PrivateKey(block.Bytes); err != nil {
+			return nil, err
+		} else if key, ok := pk.(PrivateKey); ok {
+			return key, nil
+		}
+	}
+	return nil, nil
+}
+
+// BlockToRSAPrivateKey attempts to parse a pem.Block to extract an rsa.PrivateKey
+func BlockToRSAPrivateKey(block *pem.Block) (*rsa.PrivateKey, error) {
+	if pk, err := BlockToPrivateKey(block); err != nil {
+		return nil, err
+	} else if key, ok := pk.(*rsa.PrivateKey); ok {
+		return key, nil
+	} else {
+		return nil, nil
+	}
+}
```

Comment:
This whole if masquerades as a case. IMVHO one should break it off at the first if. something like:

```go
if pk, err := BlockToPrivateKey(block); err != nil {
		return nil, err
	}
if key, ok := pk.(*rsa.PrivateKey); ok {
		return key, nil
	} 
		return nil, fmt.Errorf("not a private key")

```

Reply:
the `nil, nil` is a sort of "nothing for me to given an opinion, keep on trying" but I understand it can be confusing

Reply:
I still think you are combining unrelated checks in the dame if else if

Reply:
It's a chain... If it can decode and it's of the right type. Made via else if for readability... which I seem to have failed

Reply:
I see you apply this pattern in several places. Still not convinced but wth😕

Reply:
Do you find if arrows more readable?